control/controller_test.go

```diff
@@ -1,682 +1,225 @@
-		}
-		cancel()
-	}()
-
-	// Run a query and then wait for it to be ready.
-	q, err := ctrl.Query(context.Background(), mockCompiler)
-	if err != nil {
-		t.Fatalf("unexpected error: %s", err)
-	}
-
-	// We do not care about the results, just that the query is ready.
-	<-q.Results()
-
-	// Cancel the query. This is after the executor has already run,
-	// but before we finalize the query. This ensures that canceling
-	// at this stage will report the canceled state.
-	q.Cancel()
-
-	if want, got := Canceled, q.(*Query).State(); want != got {
-		t.Errorf("unexpected state: want=%s got=%s", want, got)
-	}
-
-	// Now finish the query by using Done.
 	q.Done()
 
-	// Verify the metrics say there are no queries.
-	gauge, err := ctrl.metrics.all.GetMetricWithLabelValues()
-	if err != nil {
-		t.Fatalf("unexpected error: %s", err)
-	}
-
-	metric := &dto.Metric{}
-	if err := gauge.Write(metric); err != nil {
-		t.Fatalf("unexpected error: %s", err)
-	}
-
-	if got, exp := int(metric.Gauge.GetValue()), 0; got != exp {
-		t.Fatalf("unexpected metric value: exp=%d got=%d", exp, got)
+	if err := q.Err(); err == nil {
+		t.Error("expected error")
+	} else if got, want := err.Error(), "expected error"; got != want {
+		t.Errorf("unexpected error -want/+got\n\t- %q\n\t+ %q", want, got)
 	}
 }
 
-func TestController_CancelQuery_Execute(t *testing.T) {
-	t.Skip("controller needs reimplementing")
-	executing := make(chan struct{})
-	defer close(executing)
-
-	executor := mock.NewExecutor()
-	executor.ExecuteFn = func(ctx context.Context, spec *plan.Spec, a *memory.Allocator) (map[string]flux.Result, <-chan flux.Metadata, error) {
-		executing <- struct{}{}
-		<-ctx.Done()
-		return nil, mock.NoMetadata, ctx.Err()
-	}
-
-	ctrl := New(Config{})
-	ctrl.executor = executor
-
-	defer func() {
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-		if err := ctrl.Shutdown(ctx); err != nil {
-			t.Fatal(err)
-		}
-		cancel()
-	}()
-
-	// Run a query and then wait for it to be ready.
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-	q, err := ctrl.Query(ctx, mockCompiler)
-	if err != nil {
-		t.Fatalf("unexpected error: %s", err)
-	}
-	<-executing
-	q.Cancel()
-
-	// We do not care about the results, just that the query is ready.
-	// We should not receive any results as the cancellation should
-	// have signaled to the executor to cancel the query.
-	select {
-	case <-q.Results():
-		// The execute function should have received the cancel signal and exited
-		// with an error.
-	case <-ctx.Done():
-		t.Error("timeout while waiting for the query to be canceled")
-	}
-	cancel()
-
-	// The state should be canceled.
-	if want, got := Canceled, q.(*Query).State(); want != got {
-		t.Errorf("unexpected state: want=%s got=%s", want, got)
-	}
-
-	// Now finish the query by using Done.
-	q.Done()
+func TestController_ShutdownWithRunningQuery(t *testing.T) {
+	ctrl := control.New(control.Config{})
+	defer shutdown(t, ctrl)
```

Comment:
do you really need this?

Reply:
The defer? It's a failsafe in case it fails somewhere else. Using shutdown multiple times should be a noop.